### DISCUSSION TITLE
Directus needs incremental migration generator to export/import CMS structure

### DISCUSSION THREAD

**Discussion by iyobo** [directus/directus#8268] [Archive]
I love this CMS...but the painful part of it now is that there is no way of synchronising my data structure accross environments or servers. I have to manually create content types and such.

I know Directus is unique in that it depends on your database, but the same way it populates a database with tables on creation (internal migrations) and reads its structure to determine new fields could easily be the same way it snapshot-deltas changes in the database.

Then that incremental snapshot delta can be saved as a migration file in one's directus repo that automatically gets run anywhere that directus repo is cloned. This would be a killer feature.

**Comment by joselcvarela** [comment]
Hello,
Can you provide an example?
It seems you can just make a dump of your db and import where you want.
Or if you want database synchronization you could a tool for your specific database. For example, for Postgres you have these: https://wiki.postgresql.org/wiki/Replication,_Clustering,_and_Connection_Pooling

**Comment by iyobo** [comment]
@joselcvarela I'm mostly thinking about the former. The thing with dumps is, it's only good for that initial deployment. It doesn't work for subsequent deployments.

Take for example, you are deploying directus from a git repo (where you have custom endpoints, etc).
You have a DEV environment where customizations are built out, and then you have a PRODUCTION environment  (with active users).

You can push **code** from dev to prod, but at that point, you cannot replace the prod DB with a dump from dev past that initial deployment. The best thing at that point is if there are migrations within the code to run. One could argue for using DB diff tools, but the problem is those are good for structural DB changes. There are also "cms config" changes that will need to be exported as well.

This is where having a good incremental migration feature would be really great as a directus core feature.
The incremental migrations can then be ran on prod with `directus bootstrap`.

**Comment by rijkvanzanten** [comment]
We just implemented schema migrations. See https://github.com/directus/directus/discussions/3891 for the full discussion on what it became, https://github.com/directus/directus/pull/7939 for the PR that implemented it, and https://docs.directus.io/reference/command-line-interface/#migrate-schema-to-a-different-environment on how to use it 🙂

**Reply by iyobo** [reply]
Yoooo! Thank you!

**Reply by joselcvarela** [reply]
Sorry @iyobo I was thinking you were asking about data migration, not schema. My bad.
Fortunately we have @rijkvanzanten, our God👼🙏, to solve this situations.

**Comment by rijkvanzanten** [comment]
This is a duplicate of https://github.com/directus/directus/discussions/3891
